alacritty_terminal/src/selection.rs

```diff
@@ -233,6 +205,24 @@ impl Selection {
         }
     }
 
+    /// Expand selection sides to include all cells.
+    pub fn include_all(&mut self) {
+        let (start, end) = (self.region.start.point, self.region.end.point);
+        let (start_side, end_side) = match self.ty {
+            SelectionType::Block
+                if start.col > end.col || (start.col == end.col && start.line < end.line) =>
+            {
+                (Side::Right, Side::Left)
+            },
+            SelectionType::Block => (Side::Left, Side::Right),
+            _ if Self::points_need_swap(start, end) => (Side::Right, Side::Left),
+            _ => (Side::Left, Side::Right),
```

Comment:
Could we merge these two match arms, it just feels strange to have `if` here.

Reply:
I don't see why this would be strange and it formats much nicer. So I'd personally like to keep it this way tbh.

Reply:
```rust
_ if Self::points_need_swap(start, end) => (Side::Right, Side::Left),
_ => (Side::Left, Side::Right),
```
vs
```rust
_ => {
    if Self::points_need_swap(start, end) {
        (Side::Right, Side::Left)
    } else {
        (Side::Left, Side::Right)
    }
}
```

So in response to "it formats much nicer", I'd argue it only formats shorter. In fact, reading the second example makes things much more clear to me because the flipped tuples are aligned.

Reply:
Looking at this whole `match` however, I really feel like it could be cleaned up to make more sense. The problem stems from a strange `points_need_swap` function, which in addition to not following it's own purpose statement, seems to only apply to non-block selections:

https://github.com/alacritty/alacritty/blob/14dc170caafc47a46d4c040a17829b52d431bbcd/alacritty_terminal/src/selection.rs#L259-L262

This requires many `mem::swap(&mut start, &mut end)` calls elsewhere. `Block` selections feel very clearly "bolted on" here, which I don't think they should.

I know I'm dredging up old debt, but I thought this might be a good chance to talk about it.

Reply:
The problem with block selections is that we cannot always swap these points in the same way.

For normal selections, we just have two points we need to potentially swap, the start and the end.

For block selection, we have four points. Because depending on logic, it makes a difference if we're talking about each corner. So the normal selection logic from start/end does not map to this. There's no way to represent the block selection in just two points like that uniformly the way it's done for the other selections. There is no simple 'needs swap' because it might just need a partial swap instead of a full one.

Making use of match guards allows keeping the nesting to only a single level, making the overhead of state one has to remember much smaller.